File: apps/mac-client/Sources/Network/StatsReporter.swift
// StatsReporter.swift — QoS statistics collector and feedback sender
// CrazyStream macOS Client
//
// Mirrors the C++ StatsReporter from crazystream-viewer. Collects real-time
// streaming statistics (packet loss, jitter, bandwidth, delay gradient) and
// periodically sends QoS feedback packets to the host every 200ms.

import Foundation

/// Snapshot of real-time streaming statistics for display in the overlay HUD.
struct StreamStats: Sendable {
    var bitrateKbps: Double = 0.0
    var fps: Double = 0.0
    var packetLoss: Double = 0.0        // 0.0 to 1.0
    var jitterMs: Double = 0.0
    var rttMs: Double = 0.0
    var codec: String = ""
    var resolutionWidth: UInt32 = 0
    var resolutionHeight: UInt32 = 0
    var connectionType: String = "p2p"
    var decodeTimeMs: Double = 0.0
    var renderTimeMs: Double = 0.0
    var framesDecoded: UInt64 = 0
    var framesDropped: UInt64 = 0
    var packetsReceived: UInt64 = 0
    var bytesReceived: UInt64 = 0

    var resolutionString: String {
        guard resolutionWidth > 0 && resolutionHeight > 0 else { return "N/A" }
        return "\(resolutionWidth)x\(resolutionHeight)"
    }
}

/// Collects streaming statistics and sends QoS feedback to the host at regular intervals.
final class StatsReporter: @unchecked Sendable {

    // MARK: - Properties

    private let transport: UDPTransport
    private weak var nackSender: NACKSender?
    private let lock = NSLock()

    /// Feedback send interval in milliseconds (200ms as specified).
    let feedbackIntervalMs: Int = 200

    // MARK: - Packet Arrival Records

    private struct PacketRecord {
        let seq: UInt16
        let senderTimestampUs: UInt32
        let recvTimeUs: UInt64
        let payloadSize: UInt32
    }

    private var recentPackets: [PacketRecord] = []
    private let maxRecentPackets = 1000

    // MARK: - Sequence Tracking (for loss calculation)

    private var expectedSeq: UInt16 = 0
    private var totalExpected: UInt64 = 0
    private var totalReceived: UInt64 = 0
    private var firstPacket = true

    // MARK: - Jitter (RFC 3550 style)

    private var jitter: Double = 0.0
    private var lastTransit: Int64 = 0
    private var jitterInitialized = false

    // MARK: - Bandwidth Estimation

    private var windowStartUs: UInt64 = 0
    private var windowBytes: UInt64 = 0

    // MARK: - Kalman Filter for Delay Gradient

    private var kalmanEstimate: Double = 0.0
    private var kalmanError: Double = 1.0
    private let kalmanQ: Double = 0.001  // Process noise
    private let kalmanR: Double = 0.1    // Measurement noise

    // MARK: - Stats (updated from external subsystems)

    private var _decodeTimeMs: Double = 0.0
    private var _renderTimeMs: Double = 0.0
    private var _codecName: String = ""
    private var _resolutionWidth: UInt32 = 0
    private var _resolutionHeight: UInt32 = 0
    private var _framesDecoded: UInt64 = 0
    private var _framesDropped: UInt64 = 0

    // MARK: - FPS Tracking

    private var fpsFrameCount: Int = 0
    private var fpsWindowStartUs: UInt64 = 0
    private var currentFps: Double = 0.0

    // MARK: - Background Timer

    private var timer: DispatchSourceTimer?
    private let timerQueue = DispatchQueue(label: "com.crazystream.stats-reporter", qos: .userInitiated)

    // MARK: - Callbacks

    /// Called each time stats are updated (on timerQueue).
    var onStatsUpdate: ((StreamStats) -> Void)?

    // MARK: - Initialization

    init(transport: UDPTransport) {
        self.transport = transport
    }

    deinit {
        stop()
    }

    /// Set the NACK sender to query for missing sequences.
    func setNackSender(_ sender: NACKSender) {
        nackSender = sender
    }

    // MARK: - Event Recording

    /// Called for each received video packet to update statistics.
    func onPacketReceived(header: VideoPacketHeader, recvTimeUs: UInt64) {
        lock.lock()
        defer { lock.unlock() }

        let seq = header.sequenceNumber
        let senderTs = header.timestampUs
        let payloadSize = header.payloadLength

        // Track packet for bandwidth / loss / jitter calculations
        let record = PacketRecord(
            seq: seq,
            senderTimestampUs: senderTs,
            recvTimeUs: recvTimeUs,
            payloadSize: payloadSize
        )
        recentPackets.append(record)
        if recentPackets.count > maxRecentPackets {
            recentPackets.removeFirst(recentPackets.count - maxRecentPackets)
        }

        // Sequence tracking for loss
        totalReceived += 1
        if firstPacket {
            expectedSeq = seq &+ 1
            firstPacket = false
            windowStartUs = recvTimeUs
            fpsWindowStartUs = recvTimeUs
        } else {
            let diff = Int16(bitPattern: seq &- expectedSeq)
            if diff >= 0 {
                totalExpected += UInt64(diff) + 1
                expectedSeq = seq &+ 1
            }
        }

        // Jitter calculation (RFC 3550 interarrival jitter)
        let transit = Int64(recvTimeUs) - Int64(senderTs)
        if jitterInitialized {
            let d = Double(abs(transit - lastTransit))
            jitter += (d - jitter) / 16.0
        } else {
            jitterInitialized = true
        }
        lastTransit = transit

        // Bandwidth window
        windowBytes += UInt64(payloadSize)

        // FPS tracking
        fpsFrameCount += 1
        let fpsDeltaUs = recvTimeUs - fpsWindowStartUs
        if fpsDeltaUs >= 1_000_000 {  // Update FPS every second
            currentFps = Double(fpsFrameCount) / (Double(fpsDeltaUs) / 1_000_000.0)
            fpsFrameCount = 0
            fpsWindowStartUs = recvTimeUs
        }
    }

    /// Update decode time from the decoder subsystem.
    func setDecodeTimeMs(_ ms: Double) {
        lock.lock()
        _decodeTimeMs = ms
        lock.unlock()
    }

    /// Update render time from the renderer subsystem.
    func setRenderTimeMs(_ ms: Double) {
        lock.lock()
        _renderTimeMs = ms
        lock.unlock()
    }

    /// Update codec name.
    func setCodecName(_ name: String) {
        lock.lock()
        _codecName = name
        lock.unlock()
    }

    /// Update resolution.
    func setResolution(width: UInt32, height: UInt32) {
        lock.lock()
        _resolutionWidth = width
        _resolutionHeight = height
        lock.unlock()
    }

    /// Increment frames decoded counter.
    func onFrameDecoded() {
        lock.lock()
        _framesDecoded += 1
        lock.unlock()
    }

    /// Increment frames dropped counter.
    func onFrameDropped() {
        lock.lock()
        _framesDropped += 1
        lock.unlock()
    }

    // MARK: - Start / Stop

    /// Start sending QoS feedback every 200ms.
    func start() {
        guard timer == nil else { return }

        let source = DispatchSource.makeTimerSource(flags: .strict, queue: timerQueue)
        source.schedule(
            deadline: .now() + .milliseconds(feedbackIntervalMs),
            repeating: .milliseconds(feedbackIntervalMs),
            leeway: .milliseconds(5)
        )
        source.setEventHandler { [weak self] in
            self?.sendFeedback()
        }
        source.resume()
        timer = source
    }

    /// Stop the feedback loop.
    func stop() {
        timer?.cancel()
        timer = nil
    }

    // MARK: - Stats Snapshot

    /// Get a snapshot of current statistics.
    func getStats() -> StreamStats {
        lock.lock()
        defer { lock.unlock() }

        return StreamStats(
            bitrateKbps: calculateBandwidthKbps(),
            fps: currentFps,
            packetLoss: calculatePacketLoss(),
            jitterMs: jitter / 1000.0,  // convert us to ms
            rttMs: 0.0,  // RTT requires round-trip measurement, not available in one-way stream
            codec: _codecName,
            resolutionWidth: _resolutionWidth,
            resolutionHeight: _resolutionHeight,
            connectionType: "p2p",
            decodeTimeMs: _decodeTimeMs,
            renderTimeMs: _renderTimeMs,
            framesDecoded: _framesDecoded,
            framesDropped: _framesDropped,
            packetsReceived: totalReceived,
            bytesReceived: windowBytes
        )
    }

    // MARK: - Feedback Sending

    /// Calculate and send a QoS feedback packet to the host.
    private func sendFeedback() {
        lock.lock()

        let lossX100 = UInt16(calculatePacketLoss() * 10000)  // 0.0250 -> 250
        let jitterUs = UInt16(min(jitter, Double(UInt16.max)))
        let bwKbps = UInt32(calculateBandwidthKbps())
        let delayGrad = calculateDelayGradientUs()
        let lastSeq = firstPacket ? 0 : expectedSeq &- 1

        lock.unlock()

        // Get missing sequences from NACK sender
        let missingSeqs = nackSender?.getMissingSequences() ?? []
        let nackCount = UInt16(min(missingSeqs.count, 64))

        // Build the QoS feedback packet
        var packet = QosFeedbackPacket(
            type: PacketType.qosFeedback.rawValue,
            flags: 0,
            lastSeqReceived: lastSeq,
            estimatedBwKbps: bwKbps,
            packetLossX100: lossX100,
            avgJitterUs: jitterUs,
            delayGradientUs: delayGrad,
            nackCount: nackCount,
            nackSeq0: missingSeqs.count > 0 ? missingSeqs[0] : 0,
            nackSeq1: missingSeqs.count > 1 ? missingSeqs[1] : 0
        )

        // Extended NACKs (beyond the first 2)
        let extendedNacks: [UInt16]
        if missingSeqs.count > 2 {
            extendedNacks = Array(missingSeqs[2..<min(missingSeqs.count, 64)])
        } else {
            extendedNacks = []
        }

        let data = packet.serialize(extendedNacks: extendedNacks)
        transport.send(data)

        // Notify stats observers
        let stats = getStats()
        onStatsUpdate?(stats)
    }

    // MARK: - Calculation Helpers

    /// Calculate packet loss rate over the tracking window.
    private func calculatePacketLoss() -> Double {
        guard totalExpected > 0 else { return 0.0 }
        let lost = totalExpected > totalReceived ? totalExpected - totalReceived : 0
        return Double(lost) / Double(totalExpected)
    }

    /// Estimate bandwidth in kbps over the recent window.
    private func calculateBandwidthKbps() -> Double {
        guard recentPackets.count >= 2 else { return 0.0 }

        let oldest = recentPackets.first!
        let newest = recentPackets.last!
        let durationUs = newest.recvTimeUs - oldest.recvTimeUs
        guard durationUs > 0 else { return 0.0 }

        var totalBytes: UInt64 = 0
        for record in recentPackets {
            totalBytes += UInt64(record.payloadSize)
        }

        let durationSec = Double(durationUs) / 1_000_000.0
        let bitsPerSec = (Double(totalBytes) * 8.0) / durationSec
        return bitsPerSec / 1000.0
    }

    /// Calculate one-way delay gradient using a simple Kalman filter.
    private func calculateDelayGradientUs() -> Int32 {
        guard recentPackets.count >= 2 else { return 0 }

        let n = recentPackets.count
        let latest = recentPackets[n - 1]
        let previous = recentPackets[n - 2]

        let transitLatest = Int64(latest.recvTimeUs) - Int64(latest.senderTimestampUs)
        let transitPrevious = Int64(previous.recvTimeUs) - Int64(previous.senderTimestampUs)
        let measurement = Double(transitLatest - transitPrevious)

        // Kalman filter update
        let prediction = kalmanEstimate
        let predictionError = kalmanError + kalmanQ

        let gain = predictionError / (predictionError + kalmanR)
        kalmanEstimate = prediction + gain * (measurement - prediction)
        kalmanError = (1.0 - gain) * predictionError

        return Int32(clamping: Int64(kalmanEstimate))
    }

    /// Current monotonic time in microseconds.
    static func currentTimeUs() -> UInt64 {
        var info = mach_timebase_info_data_t()
        mach_timebase_info(&info)
        let nanos = mach_absolute_time() * UInt64(info.numer) / UInt64(info.denom)
        return nanos / 1000
    }
}

File: libs/gridstreamer-viewer/src/qos/stats_reporter.cpp
///////////////////////////////////////////////////////////////////////////////
// stats_reporter.cpp -- QoS statistics collector and feedback sender
//
// Collects streaming statistics and sends QoS feedback to the host
// every 200ms so it can adapt encoding parameters (bitrate, resolution,
// keyframe interval) based on network conditions.
///////////////////////////////////////////////////////////////////////////////

#include "stats_reporter.h"

#include "../transport/nack_sender.h"

#include <cs/common.h>
#include <cs/transport/packet.h>

#include <chrono>
#include <cstring>
#include <cmath>
#include <algorithm>
#include <numeric>

namespace cs {

// ---------------------------------------------------------------------------
// Constructor / Destructor
// ---------------------------------------------------------------------------

StatsReporter::StatsReporter() = default;

StatsReporter::~StatsReporter() {
    stop();
}

// ---------------------------------------------------------------------------
// initialize
// ---------------------------------------------------------------------------

bool StatsReporter::initialize(int socket_fd, const ::sockaddr* peer, int peer_len) {
    std::lock_guard<std::mutex> lock(mutex_);

    socket_fd_ = socket_fd;

    if (peer && peer_len > 0) {
        peer_addr_.assign(reinterpret_cast<const uint8_t*>(peer),
                          reinterpret_cast<const uint8_t*>(peer) + peer_len);
        peer_addr_len_ = peer_len;
    }

    CS_LOG(INFO, "StatsReporter: initialized");
    return true;
}

// ---------------------------------------------------------------------------
// setNackSender
// ---------------------------------------------------------------------------

void StatsReporter::setNackSender(NackSender* nack_sender) {
    std::lock_guard<std::mutex> lock(mutex_);
    nack_sender_ = nack_sender;
}

// ---------------------------------------------------------------------------
// onPacketReceived
// ---------------------------------------------------------------------------

void StatsReporter::onPacketReceived(const VideoPacketHeader& header, uint64_t recv_time_us) {
    std::lock_guard<std::mutex> lock(mutex_);

    PacketRecord record;
    record.seq = header.sequence_number;
    record.sender_timestamp_us = header.timestamp_us;
    record.recv_time_us = recv_time_us;
    record.payload_size = header.payload_length + static_cast<uint32_t>(sizeof(VideoPacketHeader));

    // Track sequence numbers for packet loss
    if (first_packet_) {
        expected_seq_ = header.sequence_number;
        window_start_us_ = recv_time_us;
        first_packet_ = false;
    }

    total_received_++;

    // Count expected packets (handling wraparound)
    int16_t delta = static_cast<int16_t>(header.sequence_number - expected_seq_);
    if (delta > 0) {
        total_expected_ += static_cast<uint64_t>(delta);
        expected_seq_ = header.sequence_number + 1;
    } else if (delta == 0) {
        total_expected_++;
        expected_seq_++;
    }
    // Negative delta = reordered/retransmitted packet, already counted

    // Jitter calculation (RFC 3550 interarrival jitter)
    if (jitter_initialized_) {
        int64_t transit = static_cast<int64_t>(recv_time_us) -
                          static_cast<int64_t>(header.timestamp_us);
        int64_t d = transit - last_transit_;
        if (d < 0) d = -d;
        jitter_ += (static_cast<double>(d) - jitter_) / 16.0;
        last_transit_ = transit;
    } else {
        last_transit_ = static_cast<int64_t>(recv_time_us) -
                        static_cast<int64_t>(header.timestamp_us);
        jitter_initialized_ = true;
    }

    // Bandwidth tracking
    window_bytes_ += record.payload_size;

    // Store record
    recent_packets_.push_back(record);
    while (recent_packets_.size() > MAX_RECENT_PACKETS) {
        recent_packets_.pop_front();
    }
}

// ---------------------------------------------------------------------------
// start
// ---------------------------------------------------------------------------

void StatsReporter::start() {
    if (running_.load()) return;

    running_.store(true);
    feedback_thread_ = std::thread(&StatsReporter::feedbackLoop, this);

    CS_LOG(INFO, "StatsReporter: started (200ms feedback interval)");
}

// ---------------------------------------------------------------------------
// stop
// ---------------------------------------------------------------------------

void StatsReporter::stop() {
    running_.store(false);

    if (feedback_thread_.joinable()) {
        feedback_thread_.join();
    }

    CS_LOG(INFO, "StatsReporter: stopped");
}

// ---------------------------------------------------------------------------
// getStats
// ---------------------------------------------------------------------------

ViewerStats StatsReporter::getStats() const {
    std::lock_guard<std::mutex> lock(mutex_);
    std::lock_guard<std::mutex> slock(stats_mutex_);

    ViewerStats stats;
    stats.bitrate_kbps = calculateBandwidthKbps();
    stats.packet_loss = calculatePacketLoss();
    stats.jitter_ms = calculateJitterUs() / 1000.0;
    stats.decode_time_ms = decode_time_ms_;
    stats.render_time_ms = render_time_ms_;
    stats.codec = codec_name_;
    stats.resolution_width = resolution_width_;
    stats.resolution_height = resolution_height_;
    stats.frames_decoded = frames_decoded_.load();
    stats.frames_dropped = frames_dropped_.load();
    stats.packets_received = total_received_;
    stats.bytes_received = window_bytes_;

    // Estimate FPS from recent frame timestamps
    if (recent_packets_.size() >= 2) {
        uint64_t time_span = recent_packets_.back().recv_time_us -
                             recent_packets_.front().recv_time_us;
        if (time_span > 0) {
            double seconds = static_cast<double>(time_span) / 1e6;
            stats.fps = static_cast<double>(frames_decoded_.load()) > 0 ?
                        static_cast<double>(frames_decoded_.load()) / seconds : 0.0;
            // Clamp to reasonable range
            if (stats.fps > 300.0) stats.fps = 0.0;
        }
    }

    stats.connection_type = "p2p";

    return stats;
}

// ---------------------------------------------------------------------------
// Stat update methods
// ---------------------------------------------------------------------------

void StatsReporter::setDecodeTimeMs(double ms) {
    std::lock_guard<std::mutex> lock(stats_mutex_);
    decode_time_ms_ = ms;
}

void StatsReporter::setRenderTimeMs(double ms) {
    std::lock_guard<std::mutex> lock(stats_mutex_);
    render_time_ms_ = ms;
}

void StatsReporter::setCodecName(const std::string& name) {
    std::lock_guard<std::mutex> lock(stats_mutex_);
    codec_name_ = name;
}

void StatsReporter::setResolution(uint32_t width, uint32_t height) {
    std::lock_guard<std::mutex> lock(stats_mutex_);
    resolution_width_ = width;
    resolution_height_ = height;
}

void StatsReporter::onFrameDecoded() {
    frames_decoded_.fetch_add(1);
}

void StatsReporter::onFrameDropped() {
    frames_dropped_.fetch_add(1);
}

// ---------------------------------------------------------------------------
// feedbackLoop
// ---------------------------------------------------------------------------

void StatsReporter::feedbackLoop() {
    while (running_.load()) {
        std::this_thread::sleep_for(std::chrono::milliseconds(200));

        if (!running_.load()) break;

        sendFeedback();
    }
}

// ---------------------------------------------------------------------------
// sendFeedback
// ---------------------------------------------------------------------------

void StatsReporter::sendFeedback() {
    std::lock_guard<std::mutex> lock(mutex_);

    if (socket_fd_ < 0 || peer_addr_.empty()) {
        return;
    }

    QosFeedbackPacket feedback = {};
    feedback.type = static_cast<uint8_t>(PacketType::QOS_FEEDBACK);
    feedback.flags = 0;

    // Fill in stats
    if (!recent_packets_.empty()) {
        feedback.last_seq_received = recent_packets_.back().seq;
    }

    feedback.estimated_bw_kbps = static_cast<uint32_t>(calculateBandwidthKbps());

    double loss = calculatePacketLoss();
    feedback.packet_loss_x100 = static_cast<uint16_t>(loss * 10000.0);

    double jitter_us = calculateJitterUs();
    feedback.avg_jitter_us = static_cast<uint16_t>(
        std::min(jitter_us, static_cast<double>(UINT16_MAX)));

    feedback.delay_gradient_us = calculateDelayGradientUs();

    // Include NACK sequences from the NACK sender
    feedback.nack_count = 0;
    feedback.nack_seq_0 = 0;
    feedback.nack_seq_1 = 0;

    if (nack_sender_) {
        auto missing = nack_sender_->getMissingSequences();
        feedback.nack_count = static_cast<uint16_t>(std::min(missing.size(), static_cast<size_t>(UINT16_MAX)));
        if (missing.size() >= 1) feedback.nack_seq_0 = missing[0];
        if (missing.size() >= 2) feedback.nack_seq_1 = missing[1];
    }

    // Serialize and send
    auto buf = feedback.serialize();

    int sent = ::sendto(socket_fd_,
                         reinterpret_cast<const char*>(buf.data()),
                         static_cast<int>(buf.size()),
                         0,
                         reinterpret_cast<const ::sockaddr*>(peer_addr_.data()),
                         peer_addr_len_);

    if (sent <= 0) {
        CS_LOG(WARN, "StatsReporter: sendto failed: %d", cs_socket_error());
    }
}

// ---------------------------------------------------------------------------
// calculatePacketLoss
// ---------------------------------------------------------------------------

double StatsReporter::calculatePacketLoss() const {
    // Called under lock
    if (total_expected_ == 0) return 0.0;

    uint64_t lost = (total_expected_ > total_received_) ?
                    (total_expected_ - total_received_) : 0;
    return static_cast<double>(lost) / static_cast<double>(total_expected_);
}

// ---------------------------------------------------------------------------
// calculateJitterUs
// ---------------------------------------------------------------------------

double StatsReporter::calculateJitterUs() const {
    // Called under lock
    return jitter_;
}

// ---------------------------------------------------------------------------
// calculateBandwidthKbps
// ---------------------------------------------------------------------------

double StatsReporter::calculateBandwidthKbps() const {
    // Called under lock
    if (recent_packets_.size() < 2) return 0.0;

    uint64_t time_span = recent_packets_.back().recv_time_us -
                         recent_packets_.front().recv_time_us;
    if (time_span == 0) return 0.0;

    // Sum bytes in the recent window
    uint64_t total_bytes = 0;
    for (const auto& pkt : recent_packets_) {
        total_bytes += pkt.payload_size;
    }

    double seconds = static_cast<double>(time_span) / 1e6;
    double bits = static_cast<double>(total_bytes) * 8.0;
    return bits / seconds / 1000.0;  // kbps
}

// ---------------------------------------------------------------------------
// calculateDelayGradientUs
// ---------------------------------------------------------------------------

int32_t StatsReporter::calculateDelayGradientUs() const {
    // Simple Kalman filter on one-way delay measurements
    // Called under lock

    if (recent_packets_.size() < 2) return 0;

    // Calculate inter-packet delay differences
    const auto& prev = recent_packets_[recent_packets_.size() - 2];
    const auto& curr = recent_packets_.back();

    // One-way delay: (recv_time - sender_timestamp)
    int64_t prev_delay = static_cast<int64_t>(prev.recv_time_us) -
                         static_cast<int64_t>(prev.sender_timestamp_us);
    int64_t curr_delay = static_cast<int64_t>(curr.recv_time_us) -
                         static_cast<int64_t>(curr.sender_timestamp_us);

    double measurement = static_cast<double>(curr_delay - prev_delay);

    // Kalman filter update
    double predict_error = kalman_error_ + KALMAN_Q;
    double kalman_gain = predict_error / (predict_error + KALMAN_R);
    kalman_estimate_ = kalman_estimate_ + kalman_gain * (measurement - kalman_estimate_);
    kalman_error_ = (1.0 - kalman_gain) * predict_error;

    return static_cast<int32_t>(kalman_estimate_);
}

} // namespace cs
